internal/throttler/throttler.go

```diff
@@ -0,0 +1,111 @@
+//go:generate mockgen -source throttler.go -destination ../mocks/mock_throttler.go -package mocks
+
+package throttler
+
+import (
+	"context"
+	"time"
+
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+type Throttler interface {
+	Close()
+	Throttle(context.Context)
+}
+
+type NoopThrottler struct{}
+
+var _ Throttler = (*NoopThrottler)(nil)
+
+func (r *NoopThrottler) Throttle(ctx context.Context) {
+}
+
+func (r *NoopThrottler) Close() {
+}
+
+// throttler implements a throttling mechanism that can be used to control the rate of dispatched sub problems in FGA queries.
+// Throttling will start to kick in when the dispatch count exceeds the configured dispatch threshold.
+type throttler struct {
+	ticker          *time.Ticker
+	throttlingQueue chan struct{}
+	done            chan struct{}
+}
+
+// NewDispatchThrottler constructs a throttler which can be used to control the rate of dispatched sub problems in FGA queries.
+func NewDispatchThrottler(frequency time.Duration) Throttler {
+	return newDispatchThrottler(frequency)
+}
+
+// Returns a throttler instead of Throttler for testing purpose to be used internally.
+func newDispatchThrottler(frequency time.Duration) *throttler {
+	dispatchThrottler := &throttler{
+		ticker:          time.NewTicker(frequency),
+		throttlingQueue: make(chan struct{}),
+		done:            make(chan struct{}),
+	}
+	go dispatchThrottler.runTicker()
+	return dispatchThrottler
+}
+
+var (
+	dispatchThrottlingResolverDelayMsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "dispatch_throttling_resolver_delay_ms",
+		Help:                            "Time spent waiting for dispatch throttling resolver",
+		Buckets:                         []float64{1, 3, 5, 10, 25, 50, 100, 1000, 5000}, // Milliseconds. Upper bound is config.UpstreamTimeout.
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	}, []string{"grpc_service", "grpc_method"})
```

Comment:
we want to pass in whether it is dispatch or other throttled by (i.e., datastore).  So, when we construct the throttler, we pass a name to it.

Reply:
What I meant is that it may be good to have a third dimension (throttler type) associated with it.  In this way, if check is using both dispatch throttler and datastore query throttler, we can distinguish the time spent in each.